ansible/playbooks/roles/image_registry/tasks/main.yml

```diff
@@ -33,56 +33,94 @@
-          register: running_registry_image
+        - name: Get image name of registry container
+          command: >-
+            docker ps --all --format {% raw %}'{{ .Image }}'{% endraw %} -f name={{ epiphany_registry.container_name }}
+          changed_when: false
+          register: container_image
 
-        - name: Get current running registry container names
-          command: "{% raw %}docker ps --format '{{ .Names }}' -f 'name=registry' -f 'status=running'{% endraw %}"
-          register: running_registry_containers
+        - name: Perform migration
           when:
-            - running_registry_image.stdout
-            - specification.registry_image.name != running_registry_image.stdout
-
-        - set_fact:
-            container_name: "{{ running_registry_containers.stdout }}"
-          when: running_registry_containers.changed
-
-        - block:
-            - name: Stop previous version running registry container
-              command: docker stop '{{ container_name }}'
-
-            - name: Kill previous version running registry container
-              command: docker rm '{{ container_name }}'
-          when:
-            - container_name is defined
-            - container_name | length > 0
-      when: is_upgrade_run
+            - container_image.stdout | count > 0  # container exists
+            - container_image.stdout != specification.registry_image.name
+          block:
+            - name: Inspect old version registry container
+              command: docker inspect {{ epiphany_registry.container_name }}
+              changed_when: false
+              register: inspect_registry_container
+
+            - name: Parse information on registry container
+              set_fact:
+                container_facts: "{{ inspect_registry_container.stdout | from_json | first }}"
+
+            - name: Check if named volume exists
+              command: docker volume ls -f name={{ epiphany_registry.volume_name }} --quiet
+              changed_when: false
+              register: volume_ls
+
+            - name: Create named volume if not exists
+              command: docker volume create {{ epiphany_registry.volume_name }}
+              when: volume_ls.stdout | count == 0
+
+            - name: Set facts on old version registry volume
+              set_fact:
+                old_volume: "{{ container_facts.Mounts | selectattr('Destination', '==' ,'/var/lib/registry')
+                                                       | first }}"
+
+            - name: Stop old version registry container
+              command: docker stop {{ epiphany_registry.container_name }}
+              when: container_facts.State.Status == 'running'
+
+            # There is no 'docker volume rename' command
+            - name: Move data from old volume
+              shell: mv {{ old_volume.Source }}/* {{ _new_volume_path }}
```

Comment:
Is it possible to use `command` here?

Reply:
@to-bar ?

Reply:
Initially I tried 😉 But the answer is no, because of `*`.
The `command` module just runs a command, without shell, hence no asterisk expansion.